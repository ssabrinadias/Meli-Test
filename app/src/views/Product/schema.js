import gql from "graphql-tag";

const GET_ITEM = gql`
	query Item($id: String!) {
		productDescription(id: $id) {
			description
		}
		product(id: $id) {
			autor {
				name
				lastname
			}
			item {
				id
				title
				price {
					currency
					amount
					decimal
				}
				picture
				condition
				free_shipping
				sold_quantity
			}
		}
	}
`;

export default GET_ITEM;
