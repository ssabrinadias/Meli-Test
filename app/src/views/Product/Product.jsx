import React from 'react';
import styled from 'styled-components'
import { theme } from 'styled-tools'
import { Helmet } from 'react-helmet'
import Language from "../../components/language" // check location consistency for dynamic return
import Structure from "../../modules/Structure"
import Content from "../../components/Content"
import Button from "../../components/Button"


const BoxDetail =  styled.div`
    display: flex;
    background: #FFF;
`
const Detail = styled.div`
    flex-basis: 30%;
    margin-top: 32px;
    padding-right: 32px;
    font-size: 0.7em;
    h1 {
        line-height: 1.3em;
        margin-top: 16px;
        font-size: 1.4em;
    }
`
const Image = styled.div`
    flex-basis: 60%;
    max-width: 680px;
    padding-right: 16px;
    padding-top: 32px;
    img {
        display: block;
        margin: auto;
    }
`
const Price = styled.div`
    display: flex;
    margin: 32px 0;
    font-size: 2.5em;
`
const Decimal = styled.div`
    position: relative;
    top: 6px;
    left: 3px;
    font-size: 0.5em;
`

const Description = styled.div`
    padding: 32px 0 32px 32px;
    background: #FFF;
    h2 {
        margin-bottom: 32px;
        font-size: 1.5em;
    }
    p {
        font-size: 0.9em;
        color: ${theme('colors.tertiary.main')};
    }
`

const disruptResult = ({
    productDescription,
    product,
    product:{
        item:{
            id,
            title,
            condition,
            picture,
            price,
            sold_quantity,
            free_shipping
        }
    }
}) => ({
    id,
    title,
    condition:  condition==="new"? Language.ARS.new: '',
    picture,
    price,
    product,
    productDescription,
    sold_quantity,
    free_shipping
})

const ProductView = ({data}) =>  { 
    const product = data ? disruptResult(data) : {}
    const amount = (product.price||{}).amount
    const decimal = (product.price||{}).decimal
    const title = product.title || ""
    const description = (product.productDescription||{}).description || ""
 
    const decimalDisplay = (decimal)=> {
        const decimalLength = String(decimal).length
        if (decimalLength <=1 ) return `${decimal}0`
        return Decimal
    }
    
    return (
    <Structure header >
        <Helmet>
            <title>{title} Mercado Libre</title>
            <meta name="description" content={description} />
        </Helmet>

        <Content>
            <BoxDetail>
                <Image>
                    <img src={product.picture} alt={title}/>
                </Image>
                <Detail>
                    <span>{product.condition || Language.ARS.used} - {product.sold_quantity} {Language.ARS.sold}</span>
                    <h1>{title}</h1>
                    <Price>
                        <span>{Language.currency.ARS} {amount && amount.toFixed(3)}</span>
                        <Decimal>{(decimal===0 || decimal) && decimalDisplay(decimal)}</Decimal>
                    </Price>
                    <Button 
                        onClick={()=>{}}
                        primary
                        fluid
                    >
                        {Language.ARS.button.buy}
                    </Button>
                </Detail>
            </BoxDetail>
            <Description>
                <h2>
                    {Language.ARS.description}
                </h2>
                <p>{description}</p>
            </Description>
        </Content>
    </Structure>
  );
}

export default ProductView;