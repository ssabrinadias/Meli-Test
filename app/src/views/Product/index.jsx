import React from 'react';
import {useParams} from "react-router-dom";
import { useQuery } from '@apollo/react-hooks';
import ProductView from "./Product"
import GET_ITEM from "./schema"

const Product = () => {    
    const { id } = useParams()
    const { loading, error, data } = useQuery(GET_ITEM, {
      variables: { id: id },
    });
    return <ProductView
        id={id}
        data={data}
    />
}

export default Product;