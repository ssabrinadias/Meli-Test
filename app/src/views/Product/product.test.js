import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import { ApolloClient } from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import { render, cleanup } from "react-testing-library";
import GET_ITEM from "./schema";
import Product from "./Product";

const mocks = [
	{
		request: {
			query: GET_ITEM,
			variables: {
				id: "MLA835270021"
			}
		},
		result: {
			data: {
				product: {
					item: {
						id: "MLA835270021",
						title: "Samsung Galaxy A50 Dual Sim 128 Gb Negro 4 Gb Ram",
						price: {
							amount: 33.999,
							decimal: 0,
							currency: "ARS"
						},
						picture:
							"http://mla-s1-p.mlstatic.com/709651-MLA31848305203_082019-O.jpg",
						condition: "new",
						free_shipping: null,
						state: null,
						sold_quantity: 1
					},

					autor: {
						name: "Hello",
						lastname: "World"
					}
				}
			}
		}
	}
];

const client = new ApolloClient({
	cache: new InMemoryCache(),
	link: "http://localhost:3010/graphql"
});

const component = (
	<MockedProvider mocks={mocks} addTypename={false}>
		<Product />
	</MockedProvider>
);

describe("Product detail request", () => {
	afterEach(cleanup);

	it("[snapshot] - should be able to rende", () => {
		const { container } = render(component);

		expect(container.firstChild).toMatchSnapshot();
	});
});
