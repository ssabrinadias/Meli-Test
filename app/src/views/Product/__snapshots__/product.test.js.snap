// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail request [snapshot] - should be able to rende 1`] = `
<section
  class="sc-fzozJi fKZeCJ"
>
  <header
    class="sc-AxgMl fMGNCq"
  >
    <div
      class="sc-AxhUy eqTxEv"
    >
      <div
        class="sc-Axmtr hIHYcU"
      >
        <a
          class="sc-AxheI gyZAte"
          href="/"
        >
          <img
            alt="Logo Meli"
            src="[object Object]"
          />
        </a>
        <form
          class="sc-AxirZ hvoVdz"
        >
          <input
            class="sc-AxiKw iculOt"
            name="search"
            placeholder="Nunca dejes de Buscar"
            required=""
            type="text"
            value=""
          />
          <button
            class="sc-AxhCb gisRIW"
            type="submit"
          />
        </form>
      </div>
    </div>
  </header>
  <div
    class="sc-AxhUy eqTxEv"
  >
    <ul
      class="sc-AxmLO htcQYS"
    />
  </div>
  <div
    class="sc-AxhUy eqTxEv"
  >
    <div
      class="sc-fzpans kdeZNy"
    >
      <div
        class="sc-fznyAO jGTaFR"
      >
        <img />
      </div>
      <div
        class="sc-fzplWN cbmssi"
      >
        <span>
          Usado
           - 
           
          vendidos
        </span>
        <h1 />
        <div
          class="sc-fznKkj fFMFHD"
        >
          <span>
            $
             
          </span>
          <div
            class="sc-fznZeY fbjwAo"
          />
        </div>
        <button
          class="sc-fzoLsD iufAH"
        >
          Comprar
        </button>
      </div>
    </div>
    <div
      class="sc-fzokOt hUKtpy"
    >
      <h2>
        Descripci√≥n del producto
      </h2>
      <p />
    </div>
  </div>
</section>
`;
